new-frontiers-14/frontier-station-14
# ChemMaster: limit buffer, upgradable buffer size & maximum pill dosage

## About the PR



Supersedes #2523.  Should only be merged given considerations to #2542, #2544, and #2549. 

Adds a buffer limit to ChemMasters, and adds upgrade handling to the maximum volume and pill dosage limits of the machine.  ChemMaster construction now uses manipulators and matter bins instead of capacitors - the tier of manipulators used affects the max pill dosage, and the tier of matter bins affects the overall storage capacity of the ChemMaster.

Tiers are as follows:
Tier 1: 2000u, 20u pills
Tier 2: 2250u, 22u pills
Tier 3: 2500u, 25u pills
Tier 4: 3000u, 30u pills

The maximum buffer volume is available at a glance in the buffer UI if it contains anything.  The maximum pill dosage is not immediately visible, though both quantities are shown in the upgrade examine window (shift+click, check the up arrow).

## Why / Balance


Separation of purposes versus the chem dispenser.  The chem dispenser is intended for bulk chemical storage, and assuming all of the above PRs are merged as they are at time of writing, this gives you an upper bound of *40,000* units of chemicals in *one* machine, dispensable at the touch of your fingertips.

The ChemMaster's job should be the separation of chemicals and creation of pills.  This PR throws a bone to chemists by allowing larger possible doses of chemicals through pills.  It throws a bone to scientists wanting to interact with the rest of the server by incentivizing the purchase of upgrades.

## How to test


1. Spawn a chem dispenser.
2. Spawn some chemical, splash it into the ChemMaster.
3. Check the UI, the buffer volume should be listed in the buffer panel of the window (1000u by default at time of writing).
4. Spawn more jugs of the chemical, try to overfill the ChemMaster - it should neither accept nor waste resources past its buffer limit.
5. Upgrade the matter bins in the ChemMaster with the UI open - the buffer capacity should automatically update.
6. Switch to the output tab, enter the maximum pill dosage (20u), then upgrade the manipulators in the ChemMaster.  The buttons will not update - either manually enter a higher number or reduce then increase the quantity, the maximum quantity should follow the tier list above.

## Media


Delicious, dense soda pills.
![image](https://github.com/user-attachments/assets/e8e83bfc-ffd9-43ac-98de-38a8560d0995)

A fully upgraded buffer - note the immediate "/3000" visible under the Buffer section of the window.
![image](https://github.com/user-attachments/assets/ee18db72-dc81-4042-80e6-25f390769b40)

## Requirements



## Breaking changes


**Changelog**


:cl:
- tweak: ChemMaster buffer volume is limited, starting at 1000u.  Buffer volume and max pill dosage can be upgraded.
- tweak: The ChemMaster now takes 2 matter bins and manipulators to construct, not capacitors.
